pkg/volume/csi/csi_attacher.go

```diff
@@ -224,36 +243,55 @@ func (c *csiAttacher) Detach(volumeName string, nodeName types.NodeName) error {
 func (c *csiAttacher) waitForVolumeDetachment(volumeHandle, attachID string) error {
 	glog.V(4).Info(log("probing for updates from CSI driver for [attachment.ID=%v]", attachID))
 
-	ticker := time.NewTicker(c.waitSleepTime)
-	defer ticker.Stop()
+	watcher, err := c.k8s.StorageV1alpha1().VolumeAttachments().Watch(meta.SingleObject(meta.ObjectMeta{Name: attachID}))
+	if err != nil {
```

Comment:
If `VolumeAttachment` object does not exist, does `c.k8s.StorageV1alpha1().VolumeAttachments().Watch(...)` fail with a special error? If so, we should handle that error and not fail in that case.

Reply:
no, nothing happens.

Reply:
I'm not sure if this is correct. I spoke with @lavalamp his suggestion was to do a GET operation to fetch the object. If it does not exist then we assume it is deleted. If it exists, we can use the resource ID from the returned object to then set up a watch. If the object is deleted before we start the watch, as long as we use the resource ID we will observe the delete event.

Reply:
You need to do a GET, and then sets up the watch with the resourceVersion in the GET response. The watch is guaranteed to stream the changes after that resourceVersion. Otherwise the watch is started at an arbitrary point in apiserver's cached history.

Another caveat is that if the watch is started with a specific resourceVersion, it might be already out of the window of apiserver's cache. API server will return an error like "too old resource version...". In that case, you should repeat GET + watch cycle.

Reply:
@saad-ali @caesarxuchao Fixed, Thanks for your help, PTAL!